
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Welcome to cgr_view’s documentation! &#8212; cgr_view 0.0.1dev documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="welcome-to-cgr-view-s-documentation">
<h1>Welcome to cgr_view’s documentation!<a class="headerlink" href="#welcome-to-cgr-view-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
<span class="target" id="module-cgr"></span><p>cgr</p>
<p>A module for creating, saving and drawing k-mer matrices and Chaos Game Representations (CGRs) of nucleotide sequences</p>
<div class="section" id="prerequisites">
<h2>Prerequisites<a class="headerlink" href="#prerequisites" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Jellyfish</p></li>
</ul>
<p>An external program for counting k-mers. Must be accessible on the path. You can install from conda as follows:</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda install -c bioconda jellyfish
</pre></div>
</div>
</div></blockquote>
</div>
<div class="section" id="quickstart">
<h2>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Input fasta file, get cgr</p>
<blockquote>
<div><ul class="simple">
<li><p>one cgr for each entry in the fasta file</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;my_seqs.fa&quot;</span><span class="p">,</span> <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;my_cgrs&quot;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">7</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p>just one cgr with all entries in the fasta file (eg for genomes and contigs)</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;my_genome.fa&quot;</span><span class="p">,</span> <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;genome_cgr&quot;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">as_single</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="workflow">
<h2>Workflow:<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>make kmer count db in Jellyfish from fasta -&gt; generate cgr from db.</p></li>
<li><p>optionally merge cgrs into single cgr as separate channels</p></li>
<li><p>stack all composed cgrs into an array of cgrs</p></li>
<li><p>save as numpy binary (.npy) files</p></li>
</ol>
</div>
<div class="section" id="usage">
<h2>Usage:<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><p>Import module</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">cgr</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Make kmer count db</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">run_jellyfish</span><span class="p">(</span><span class="s2">&quot;test_data/NC_012920.fasta&quot;</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="s2">&quot;11mer.jf&quot;</span><span class="p">)</span>
<span class="n">cgr</span><span class="o">.</span><span class="n">run_jellyfish</span><span class="p">(</span><span class="s2">&quot;test_data/NC_012920.fasta&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="s2">&quot;10_mer.jf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ol>
<ol class="arabic" start="2">
<li><p>Load CGRs from kmer count db</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr1</span> <span class="o">=</span> <span class="n">cgr</span><span class="o">.</span><span class="n">cgr_matrix</span><span class="p">(</span><span class="s2">&quot;/Users/macleand/Desktop/athal-5-mers.jf&quot;</span><span class="p">)</span>
<span class="n">cgr2</span> <span class="o">=</span> <span class="n">cgr</span><span class="o">.</span><span class="n">cgr_matrix</span><span class="p">(</span><span class="s2">&quot;test_data/five_mer.jf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Draw a cgr and save to file</p>
<blockquote>
<div><ul class="simple">
<li><p>just one cgr, can choose colour (value of ‘h’) and which channel to put cgr in</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">draw_cgr</span><span class="p">(</span><span class="n">cgr1</span><span class="p">,</span> <span class="n">h</span> <span class="o">=</span> <span class="mf">0.64</span><span class="p">,</span> <span class="n">v</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">out</span> <span class="o">=</span> <span class="s2">&quot;my_cgr.png&quot;</span><span class="p">,</span> <span class="n">resize</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">main</span> <span class="o">=</span> <span class="s2">&quot;s&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p>two cgrs, first in tuple goes in ‘h’, second goes in ‘s’. Can set ‘v’</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">draw_cgr</span><span class="p">(</span> <span class="p">(</span><span class="n">cgr1</span><span class="p">,</span> <span class="n">cgr1</span><span class="p">),</span> <span class="n">v</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">out</span> <span class="o">=</span> <span class="s2">&quot;two_cgrs.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p>three cgrs ‘h’,’s’ and ‘v’ are assigned as order in tuple</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">draw_cgr</span><span class="p">(</span> <span class="p">(</span><span class="n">cgr1</span><span class="p">,</span> <span class="n">cgr1</span><span class="p">,</span> <span class="n">cgr1</span><span class="p">)</span> <span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Save a single cgr into a text file</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">save_as_csv</span><span class="p">(</span><span class="n">cgr1</span><span class="p">,</span> <span class="nb">file</span> <span class="o">=</span> <span class="s2">&quot;out.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Join n cgrs into one, extending the number of channels …</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">merged_cgr</span> <span class="o">=</span> <span class="n">cgr</span><span class="o">.</span><span class="n">join_cgr</span><span class="p">(</span> <span class="p">(</span><span class="n">cgr1</span><span class="p">,</span> <span class="n">cgr2</span><span class="p">,</span> <span class="o">...</span> <span class="p">)</span> <span class="p">)</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><p>Write to file (numpy binary)</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">save_cgr</span><span class="p">(</span><span class="s2">&quot;my_cgr, merged_cgr )</span>
</pre></div>
</div>
</div></blockquote>
</li>
<li><dl>
<dt>Input fasta file, get cgr</dt><dd><ul class="simple">
<li><p>one cgr for each entry in the fasta file</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;my_seqs.fa&quot;</span><span class="p">,</span> <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;my_cgrs&quot;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">7</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p>just one cgr with all entries in the fasta file (eg for genomes and contigs)</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cgr</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="s2">&quot;my_genome.fa&quot;</span><span class="p">,</span> <span class="n">outfile</span> <span class="o">=</span> <span class="s2">&quot;genome_cgr&quot;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">as_single</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</li>
</ol>
<dl class="function">
<dt id="cgr.blocky_scale">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">blocky_scale</code><span class="sig-paren">(</span><em class="sig-param">im: numpy.ndarray</em>, <em class="sig-param">nR: int</em>, <em class="sig-param">nC: int</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#cgr.blocky_scale" title="Permalink to this definition">¶</a></dt>
<dd><p>Upscales an array in preparation for drawing. By default the array is a square with sqrt(k ** 4) wide and high.
For many values of k this will be too small to view well on a monitor. This function does a scale operartion that
increases the size of the image by simply increasing the pixels in each square.</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>im numpy.ndarray – the image to be scaled</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>nR int – the number of height pixels to be in the final image</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>nC int – the number of width pixels to be in the final image</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>numpy.ndarray – upscaled image</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.cgr_matrix">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">cgr_matrix</code><span class="sig-paren">(</span><em class="sig-param">jellyfish: str</em><span class="sig-paren">)</span> &#x2192; scipy.sparse.dok.dok_matrix<a class="headerlink" href="#cgr.cgr_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Main function, creates the cgr matrix, a sparse matrix of type scipy.sparse.dok_matrix</p>
<p>Runs the cgr process on a jellyfish file and returns a scipy.sparse.dok_matrix object of the CGR with dtype int32
Only observed kmers are represented, absent coordinates mean 0 counts for the kmer at that coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>jellyfish str – jellyfish DB file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>scipy.sparse.dok_matrix – sparse matrix of kmer counts</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.draw">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">draw</code><span class="sig-paren">(</span><em class="sig-param">rgb: numpy.ndarray</em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#cgr.draw" title="Permalink to this definition">¶</a></dt>
<dd><p>renders RGB array on the screen.</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>rgb numpy.ndarray – RGB channel image</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.draw_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">draw_cgr</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrices: scipy.sparse.dok.dok_matrix</em>, <em class="sig-param">h: float = 0.8</em>, <em class="sig-param">s: float = 0.5</em>, <em class="sig-param">v: float = 1.0</em>, <em class="sig-param">main: str = 's'</em>, <em class="sig-param">show: bool = True</em>, <em class="sig-param">write: bool = True</em>, <em class="sig-param">out: str = 'cgr.png'</em>, <em class="sig-param">resize: bool = False</em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#cgr.draw_cgr" title="Permalink to this definition">¶</a></dt>
<dd><p>Draws cgrs to a file. Allows user to set which of up to 3 provided cgr matrices goes in at which of the H, S or V image channels.
Typically for one channel, select h to specify the image colour and set cgr as s to change that colour according to counts in cgr.
Set v to 1.0 for maximum brightness.</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgr_matrices scipy.sparse.dok_matrix or tuple of scipy.sparse.dok_matrix elements, cgrs to be drawn. Tuple provides order for HSV channels of image.</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>h float – (0..1) value for h channel if not used for cgr data</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>s float – (0..1) value for s channel if not used for cgr data</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>v float – (0..1) value for v channel if not used for cgr data</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>main str – the channel to place the cgr matrix in if a single cgr matrix is passed</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>show bool – render CGR picture to screen</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>write – write CGR picture to file</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>out str – filename to write to</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>resize bool or int – if False no image resizing is done, if an int image is rescaled to resize pixels width and height</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.draw_single_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">draw_single_cgr</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrix</em>, <em class="sig-param">h=0.8</em>, <em class="sig-param">s=0.5</em>, <em class="sig-param">v=1.0</em>, <em class="sig-param">main='s'</em>, <em class="sig-param">show=True</em>, <em class="sig-param">write=True</em>, <em class="sig-param">out='cgr.png'</em>, <em class="sig-param">resize=False</em><span class="sig-paren">)</span><a class="headerlink" href="#cgr.draw_single_cgr" title="Permalink to this definition">¶</a></dt>
<dd><p>draws a single cgr image, selecting channels and resizing as appropriate</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgr_matrix scipy.sparse.dok_matrix to be drawn.</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>h float – (0..1) value for h channel if not used for cgr data</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>s float – (0..1) value for s channel if not used for cgr data</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>v float – (0..1) value for v channel if not used for cgr data</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>main str – the channel to place the cgr matrix in</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>show bool – render CGR picture to screen</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>write – write CGR picture to file</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>out str – filename to write to</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>resize bool or int – if False no image resizing is done, if an int image is rescaled to resize pixels width and height</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.draw_three_cgrs">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">draw_three_cgrs</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrices</em>, <em class="sig-param">show=True</em>, <em class="sig-param">write=True</em>, <em class="sig-param">out='cgr.png'</em>, <em class="sig-param">resize=False</em><span class="sig-paren">)</span><a class="headerlink" href="#cgr.draw_three_cgrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Draws a tuple of 3 cgr matrices as an image</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgr_matrices tuple of scipy.sparse.dok_matrix elements, cgrs to be drawn. Tuple provides order for HSV channels of image</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>show bool – render CGR picture to screen</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>write – write CGR picture to file</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>out str – filename to write to</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>resize bool or int – if False no image resizing is done, if an int image is rescaled to resize pixels width and height</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.draw_two_cgrs">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">draw_two_cgrs</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrices</em>, <em class="sig-param">v=1.0</em>, <em class="sig-param">show=True</em>, <em class="sig-param">write=True</em>, <em class="sig-param">out='cgr.png'</em>, <em class="sig-param">resize=False</em><span class="sig-paren">)</span><a class="headerlink" href="#cgr.draw_two_cgrs" title="Permalink to this definition">¶</a></dt>
<dd><p>draws two cgr matrices into a single image. first matrix of tuple becomes h channel, second of tuple becomes v channel</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgr_matrices  tuple of scipy.sparse.dok_matrix elements, cgrs to be drawn.</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>v float – (0..1) value for v channel</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>show bool – render CGR picture to screen</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>write – write CGR picture to file</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>out str – filename to write to</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>resize bool or int – if False no image resizing is done, if an int image is rescaled to resize pixels width and height</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.estimate_genome_size">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">estimate_genome_size</code><span class="sig-paren">(</span><em class="sig-param">fasta: str</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#cgr.estimate_genome_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Guesses genome size from fasta file size, assumes 1 byte ~= 1 base</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>fasta str – a fasta file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int – approximate genome size in nucleotides</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.from_fasta">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">from_fasta</code><span class="sig-paren">(</span><em class="sig-param">fasta_file: str</em>, <em class="sig-param">outfile: str = 'my_cgrs'</em>, <em class="sig-param">as_single: bool = False</em>, <em class="sig-param">k: int = 7</em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#cgr.from_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Factory function to load in a FASTA file and generate a binary .npy of CGRs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta_file</strong> – str FASTA file to load</p></li>
<li><p><strong>outfile</strong> – str outfile to save</p></li>
<li><p><strong>as_single</strong> – bool If True treats all entries as single sequence and return one CGR. If False, treats all entries individually and returns many CGR</p></li>
<li><p><strong>k</strong> – int length of kmer to use</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.get_coord">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">get_coord</code><span class="sig-paren">(</span><em class="sig-param">kmer: str</em><span class="sig-paren">)</span> &#x2192; List[int]<a class="headerlink" href="#cgr.get_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>given a kmer gets the coordinates of the box position in the cgr grid,
returns as list [x,y] of coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>kmer str – a string of nucleotides</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>coords [x,y] – the x,y positions of the nucleotides in the cgr</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.get_grid_size">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">get_grid_size</code><span class="sig-paren">(</span><em class="sig-param">k: int</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#cgr.get_grid_size" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the grid size (total number of elements for a
cgr of k length kmers</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>k int – the value of k to be used</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int – the total number of elements in the grid</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.get_k">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">get_k</code><span class="sig-paren">(</span><em class="sig-param">jellyfish: str</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#cgr.get_k" title="Permalink to this definition">¶</a></dt>
<dd><p>asks the jellyfish file what value was used for k</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>jellyfish str – jellyfish DB file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int – length of k used</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.get_kmer_list">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">get_kmer_list</code><span class="sig-paren">(</span><em class="sig-param">jellyfish: str</em><span class="sig-paren">)</span> &#x2192; Generator[List, str, None]<a class="headerlink" href="#cgr.get_kmer_list" title="Permalink to this definition">¶</a></dt>
<dd><p>runs jellyfish dump on a Jellyfish DB. Captures output as a generator stream.
Each item returned is a list [kmer: str, count: str]</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>jellyfish str – a Jellyfish DB file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Generator – a list of [kmer string, times_kmer_seen]</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.is_cgr_matrix">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">is_cgr_matrix</code><span class="sig-paren">(</span><em class="sig-param">obj</em><span class="sig-paren">)</span> &#x2192; bool<a class="headerlink" href="#cgr.is_cgr_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>returns true if obj is a scipy.sparse.dok.dok_matrix object</p>
</dd></dl>

<dl class="function">
<dt id="cgr.join_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">join_cgr</code><span class="sig-paren">(</span><em class="sig-param">cgrs: tuple</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#cgr.join_cgr" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes tuple of cgrs of shape (n,n) and returns one stacked array of size (n,n, len(cgrs) )</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgrs tuple – tuple of cgrs to be joined</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.load_npy">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">load_npy</code><span class="sig-paren">(</span><em class="sig-param">file: str</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#cgr.load_npy" title="Permalink to this definition">¶</a></dt>
<dd><p>loads numpy .npy file as ndarray.
Useful for restoring collections of cgrs but resulting array is not compatible directly with
drawing methods here.</p>
<p>:param file str – numpy .npy file to load
:return: numpy.ndarray</p>
</dd></dl>

<dl class="function">
<dt id="cgr.make_blanks_like">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">make_blanks_like</code><span class="sig-paren">(</span><em class="sig-param">a: scipy.sparse.dok.dok_matrix</em>, <em class="sig-param">h: float = 1.0</em>, <em class="sig-param">s: float = 1.0</em>, <em class="sig-param">v: float = 1.0</em><span class="sig-paren">)</span> &#x2192; Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray]<a class="headerlink" href="#cgr.make_blanks_like" title="Permalink to this definition">¶</a></dt>
<dd><p>returns tuple of numpy.ndarrays with default values of h,s and v of shape of a</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>a scipy.sparse.dok_matrix – a cgr matrix to make blanks like</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>h float – the values with which to fill the first numpy.ndarray</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>s float – the values with which to fill the second numpy.ndarray</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>v float – the values with which to fill the third numpy.ndarray</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Tuple of numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.many_seq_record_to_many_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">many_seq_record_to_many_cgr</code><span class="sig-paren">(</span><em class="sig-param">seq_record: &lt;module 'Bio.SeqIO.FastaIO' from '/Users/macleand/miniconda2/envs/cgr_view/lib/python3.7/site-packages/Bio/SeqIO/FastaIO.py'&gt;</em>, <em class="sig-param">k: int</em><span class="sig-paren">)</span> &#x2192; scipy.sparse.dok.dok_matrix<a class="headerlink" href="#cgr.many_seq_record_to_many_cgr" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq_record</strong> – Bio.SeqIO FASTA record</p></li>
<li><p><strong>k</strong> – int size of k to use</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>scipy.sparse.dok_matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.many_seq_record_to_one_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">many_seq_record_to_one_cgr</code><span class="sig-paren">(</span><em class="sig-param">fa_file: str</em>, <em class="sig-param">k: int</em><span class="sig-paren">)</span> &#x2192; scipy.sparse.dok.dok_matrix<a class="headerlink" href="#cgr.many_seq_record_to_one_cgr" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads many sequence records in a FASTA file into a single CGR matrix, treating all sequence records as if they are one sequence, EG as if for a genome sequence in Chromosomes.
:param fa_file: str FASTA FILE name
:param k: int length of k to use
:return: scipy.sparse.dok_matrix</p>
</dd></dl>

<dl class="function">
<dt id="cgr.resize_rgb_out">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">resize_rgb_out</code><span class="sig-paren">(</span><em class="sig-param">rgb: numpy.ndarray</em>, <em class="sig-param">resize: int</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#cgr.resize_rgb_out" title="Permalink to this definition">¶</a></dt>
<dd><p>given an rgb image in one pixel per kmer size, increases size so that the resulting image is resize * resize pixels</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>rgb numpy.ndarray – an RGB image array</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>resize – pixel width (and therefore height) of resulting image</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>numpy.ndarray – resized image with shape (resize, resize)</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.run_jellyfish">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">run_jellyfish</code><span class="sig-paren">(</span><em class="sig-param">fasta: str</em>, <em class="sig-param">k: int</em>, <em class="sig-param">out: str</em><span class="sig-paren">)</span> &#x2192; int<a class="headerlink" href="#cgr.run_jellyfish" title="Permalink to this definition">¶</a></dt>
<dd><p>runs Jellyfish on fasta file using k kmer size, produces Jellyfish db file as side effect.</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>fasta str – a fasta file</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>k int – size of kmers to use</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>out str – file in which to save kmer db</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int – return code of Jellyfish subprocess</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.save_as_csv">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">save_as_csv</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrix: scipy.sparse.dok.dok_matrix</em>, <em class="sig-param">file: str = 'cgr_matrix.csv'</em>, <em class="sig-param">delimiter: str = '</em>, <em class="sig-param">'</em>, <em class="sig-param">fmt: str = '%d'</em><span class="sig-paren">)</span><a class="headerlink" href="#cgr.save_as_csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes simple 1 channel cgr matrix to CSV file.</p>
<p>See also numpy.savetxt</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgr_matrix scipy.sparse.dok_matrix – cgr_matrix to save</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>file str – filename to write to</p>
</dd>
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>delimiter str – column separator character</p>
</dd>
<dt class="field-even">Param</dt>
<dd class="field-even"><p>fmt str – text format string</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.save_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">save_cgr</code><span class="sig-paren">(</span><em class="sig-param">cgr_obj: numpy.ndarray</em>, <em class="sig-param">outfile: str = 'cgr'</em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#cgr.save_cgr" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves cgr_obj as numpy .npy file.
cgr_obj one or more dimensional numpy.ndarray.
saves as ndarray not dokmatrix, so can be loaded in regular numpy as collections of cgrs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cgr_obj</strong> – numpy.ndarray constructed cgr_object to save</p></li>
<li><p><strong>outfile</strong> – str file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.scale_cgr">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">scale_cgr</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrix: scipy.sparse.dok.dok_matrix</em><span class="sig-paren">)</span> &#x2192; scipy.sparse.dok.dok_matrix<a class="headerlink" href="#cgr.scale_cgr" title="Permalink to this definition">¶</a></dt>
<dd><p>returns scaled version of cgr_matrix in range 0..1</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>cgr_matrix scipy.sparse.dok_matrix – matrix to scale</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>scaled scipy.sparse.dok_matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.stack_cgrs">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">stack_cgrs</code><span class="sig-paren">(</span><em class="sig-param">cgr_matrices: Tuple</em><span class="sig-paren">)</span> &#x2192; numpy.ndarray<a class="headerlink" href="#cgr.stack_cgrs" title="Permalink to this definition">¶</a></dt>
<dd><p>stacks cgrs of tuple of N numpy.ndarrays of shape (w,h)
returns ndarray of ndarrays of shape (w,h,N)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cgr_matrices</strong> – tuple of cgr_matrices</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="cgr.write_out">
<code class="sig-prename descclassname">cgr.</code><code class="sig-name descname">write_out</code><span class="sig-paren">(</span><em class="sig-param">rgb: numpy.ndarray</em>, <em class="sig-param">out: str</em>, <em class="sig-param">resize: int</em><span class="sig-paren">)</span> &#x2192; None<a class="headerlink" href="#cgr.write_out" title="Permalink to this definition">¶</a></dt>
<dd><p>writes RGB array as image</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rgb</strong> – numpy.ndarray – RGB channel image</p></li>
<li><p><strong>out</strong> – str file to write to</p></li>
<li><p><strong>resize</strong> – bool or int. If False will not resize, if int will resize image up to that size</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">cgr_view</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Dan MacLean.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>